<a href="https://colab.research.google.com/github/ivashenceva220495-beep/My_first_repository/blob/main/statistika_dz1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
#Подгружаю копию файла с названием столбцов из файла horses_data.names
df = pd.read_csv('horse_data_copy.csv', na_values = "?")
df.head()

,surgery,age,hospital number,rectal temperature,pulse,respiratory rate,temperature of extremities,peripheral pulse,mucous membranes,capillary refill time,...,packed cell volume,total protein,abdominocentesis appearance,abdomcentesis total protein,outcome,surgical lesion?,type of lesion,type of lesion.1,type of lesion.2,cp_data
0,2.0,1,530101,38.5,66.0,28.0,3.0,3.0,NaN,2.0,...,45.0,8.4,NaN,NaN,2.0,2,11300,0,0,2
1,1.0,1,534817,39.2,88.0,20.0,NaN,NaN,4.0,1.0,...,50.0,85.0,2.0,2.0,3.0,2,2208,0,0,2
2,2.0,1,530334,38.3,40.0,24.0,1.0,1.0,3.0,1.0,...,33.0,6.7,NaN,NaN,1.0,2,0,0,0,1
3,1.0,9,5290409,39.1,164.0,84.0,4.0,1.0,6.0,2.0,...,48.0,7.2,3.0,5.3,2.0,1,2208,0,0,1
4,2.0,1,530255,37.3,104.0,35.0,NaN,NaN,6.0,2.0,...,74.0,7.4,NaN,NaN,2.0,2,4300,0,0,2


Задание 1
Изучить представленный набор данных на основе описания его столбцов, загрузить его и оставить 8 столбцов для дальнейшего изучения: surgery?, Age, rectal temperature, pulse, respiratory rate, temperature of extremities, pain, outcome.

In [ ]:
file_ = open('horse_data.names')
print(file_.read())

1. TItle: Horse Colic database

2. Source Information
   -- Creators: Mary McLeish & Matt Cecile
	  	Department of Computer Science
		University of Guelph
		Guelph, Ontario, Canada N1G 2W1
		mdmcleish@water.waterloo.edu
   -- Donor:    Will Taylor (taylor@pluto.arc.nasa.gov)
   -- Date:     8/6/89

3. Past Usage:
   -- Unknown

4. Relevant Information:

   -- 2 data files
      -- horse-colic.data: 300 training instances
      -- horse-colic.test: 68 test instances
   -- Possible class attributes: 24 (whether lesion is surgical)
     -- others include: 23, 25, 26, and 27
   -- Many Data types: (continuous, discrete, and nominal)

5. Number of Instances: 368 (300 for training, 68 for testing)

6. Number of attributes: 28

7. Attribute Information:

  1:  surgery?
          1 = Yes, it had surgery
          2 = It was treated without surgery

  2:  Age
          1 = Adult horse
          2 = Young (< 6 months)

  3:  Hospital Number
          - numeric id
          - the case number assi

In [ ]:
#выводим необходимые столбцы
df = df[["surgery","age","rectal temperature","pulse","respiratory rate","temperature of extremities","pain","outcome"]]
df

,surgery,age,rectal temperature,pulse,respiratory rate,temperature of extremities,pain,outcome
0,2.0,1,38.5,66.0,28.0,3.0,5.0,2.0
1,1.0,1,39.2,88.0,20.0,NaN,3.0,3.0
2,2.0,1,38.3,40.0,24.0,1.0,3.0,1.0
3,1.0,9,39.1,164.0,84.0,4.0,2.0,2.0
4,2.0,1,37.3,104.0,35.0,NaN,NaN,2.0
...,...,...,...,...,...,...,...,...
295,1.0,1,NaN,120.0,70.0,4.0,2.0,3.0
296,2.0,1,37.2,72.0,24.0,3.0,4.0,3.0
297,1.0,1,37.5,72.0,30.0,4.0,4.0,2.0
298,1.0,1,36.5,100.0,24.0,3.0,3.0,1.0


Задание 2
Проанализировать значения по столбцам, рассчитать базовые статистики, найти выбросы.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery                     299 non-null    float64
 1   age                         300 non-null    int64  
 2   rectal temperature          240 non-null    float64
 3   pulse                       276 non-null    float64
 4   respiratory rate            242 non-null    float64
 5   temperature of extremities  244 non-null    float64
 6   pain                        245 non-null    float64
 7   outcome                     299 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.9 KB


In [ ]:
#находим пропуски
(df.isna().mean() * 100).round(2)

surgery                        0.33
age                            0.00
rectal temperature            20.00
pulse                          8.00
respiratory rate              19.33
temperature of extremities    18.67
pain                          18.33
outcome                        0.33
dtype: float64

In [ ]:
#находим базовые статистические данные для столбца "age"
print(df['age'].mean())
print(df['age'].median())
print(df['age'].mode()[0])

print(df['age'].std())
print(df['age'].var())

1.64
1.0
1
2.173971905557624
4.726153846153846


In [ ]:
#находим базовые статистические данные для столбца "surgery"
print(df['surgery'].mean())
print(df['surgery'].median())
print(df['surgery'].mode()[0])

print(df['surgery'].std())
print(df['surgery'].var())

1.3979933110367893
1.0
1.0
0.4903046443288189
0.24039864425040958


In [ ]:
#находим базовые статистические данные для столбца "rectal temperature"
print(df['rectal temperature'].mean())
print(df['rectal temperature'].median())
print(df['rectal temperature'].mode()[0])

print(df['rectal temperature'].std())
print(df['rectal temperature'].var())

38.16791666666667
38.2
38.0
0.7322886641121578
0.5362466875871686


In [ ]:
#находим базовые статистические данные для столбца "pulse"
print(df['pulse'].mean())
print(df['pulse'].median())
print(df['pulse'].mode()[0])

print(df['pulse'].std())
print(df['pulse'].var())

71.91304347826087
64.0
48.0
28.630556660735
819.7087747035573


In [ ]:
#находим базовые статистические данные для столбца "respiratory rate"
print(df['respiratory rate'].mean())
print(df['respiratory rate'].median())
print(df['respiratory rate'].mode()[0])

print(df['respiratory rate'].std())
print(df['respiratory rate'].var())

30.417355371900825
24.5
20.0
17.64223138513467
311.2483282466308


In [ ]:
#находим базовые статистические данные для столбца "temperature of extremities"
print(df['temperature of extremities'].mean())
print(df['temperature of extremities'].median())
print(df['temperature of extremities'].mode()[0])

print(df['temperature of extremities'].std())
print(df['temperature of extremities'].var())

2.348360655737705
3.0
3.0
1.0450536920112763
1.0921372191863994


In [ ]:
#находим базовые статистические данные для столбца "pain"
print(df['pain'].mean())
print(df['pain'].median())
print(df['pain'].mode()[0])

print(df['pain'].std())
print(df['pain'].var())

2.951020408163265
3.0
3.0
1.3079395711243496
1.7107059217129477


In [ ]:
#находим базовые статистические данные для столбца "outcome"
print(df['outcome'].mean())
print(df['outcome'].median())
print(df['outcome'].mode()[0])

print(df['outcome'].std())
print(df['outcome'].var())

1.5518394648829432
1.0
1.0
0.7371869448054943
0.5434445915916589


In [ ]:
#находим выбросы через межквартальный размах
q1 = df['age'].quantile(0.25)
q3 = df['age'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
remove_outliers = df[df['age'].between(lower_bound, upper_bound)]
remove_outliers

,surgery,age,rectal temperature,pulse,respiratory rate,temperature of extremities,pain,outcome
0,2.0,1,38.5,66.0,28.0,3.0,5.0,2.0
1,1.0,1,39.2,88.0,20.0,NaN,3.0,3.0
2,2.0,1,38.3,40.0,24.0,1.0,3.0,1.0
4,2.0,1,37.3,104.0,35.0,NaN,NaN,2.0
5,2.0,1,NaN,NaN,NaN,2.0,2.0,1.0
...,...,...,...,...,...,...,...,...
295,1.0,1,NaN,120.0,70.0,4.0,2.0,3.0
296,2.0,1,37.2,72.0,24.0,3.0,4.0,3.0
297,1.0,1,37.5,72.0,30.0,4.0,4.0,2.0
298,1.0,1,36.5,100.0,24.0,3.0,3.0,1.0


In [ ]:
# выбросы для столбца age
df[~df['age'].between(lower_bound, upper_bound)]

,surgery,age,rectal temperature,pulse,respiratory rate,temperature of extremities,pain,outcome
3,1.0,9,39.1,164.0,84.0,4.0,2.0,2.0
9,2.0,9,38.3,90.0,NaN,1.0,5.0,1.0
13,2.0,9,38.0,92.0,28.0,1.0,1.0,2.0
16,1.0,9,NaN,128.0,36.0,3.0,4.0,2.0
23,1.0,9,38.3,130.0,60.0,NaN,2.0,1.0
39,1.0,9,39.2,146.0,96.0,NaN,NaN,2.0
41,2.0,9,39.0,150.0,72.0,NaN,NaN,1.0
55,1.0,9,38.6,160.0,20.0,3.0,3.0,2.0
74,1.0,9,NaN,NaN,NaN,NaN,NaN,2.0
75,1.0,9,39.7,100.0,NaN,3.0,2.0,3.0


In [ ]:
# выбросы для столбца age
df[~df['surgery'].between(lower_bound, upper_bound)]

,surgery,age,rectal temperature,pulse,respiratory rate,temperature of extremities,pain,outcome
132,NaN,1,38.0,48.0,20.0,3.0,4.0,NaN


In [ ]:
#находим выбросы через межквартальный размах
q1 = df['surgery'].quantile(0.25)
q3 = df['surgery'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
remove_outliers = df[df['surgery'].between(lower_bound, upper_bound)]
remove_outliers

,surgery,age,rectal temperature,pulse,respiratory rate,temperature of extremities,pain,outcome
0,2.0,1,38.5,66.0,28.0,3.0,5.0,2.0
1,1.0,1,39.2,88.0,20.0,NaN,3.0,3.0
2,2.0,1,38.3,40.0,24.0,1.0,3.0,1.0
3,1.0,9,39.1,164.0,84.0,4.0,2.0,2.0
4,2.0,1,37.3,104.0,35.0,NaN,NaN,2.0
...,...,...,...,...,...,...,...,...
295,1.0,1,NaN,120.0,70.0,4.0,2.0,3.0
296,2.0,1,37.2,72.0,24.0,3.0,4.0,3.0
297,1.0,1,37.5,72.0,30.0,4.0,4.0,2.0
298,1.0,1,36.5,100.0,24.0,3.0,3.0,1.0


In [ ]:
q1 = df['rectal temperature'].quantile(0.25)
q3 = df['rectal temperature'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
remove_outliers = df[df['rectal temperature'].between(lower_bound, upper_bound)]
remove_outliers

,surgery,age,rectal temperature,pulse,respiratory rate,temperature of extremities,pain,outcome
0,2.0,1,38.5,66.0,28.0,3.0,5.0,2.0
1,1.0,1,39.2,88.0,20.0,NaN,3.0,3.0
2,2.0,1,38.3,40.0,24.0,1.0,3.0,1.0
3,1.0,9,39.1,164.0,84.0,4.0,2.0,2.0
4,2.0,1,37.3,104.0,35.0,NaN,NaN,2.0
...,...,...,...,...,...,...,...,...
292,1.0,1,37.0,66.0,20.0,1.0,4.0,2.0
294,2.0,1,38.5,40.0,16.0,1.0,2.0,1.0
296,2.0,1,37.2,72.0,24.0,3.0,4.0,3.0
297,1.0,1,37.5,72.0,30.0,4.0,4.0,2.0


In [ ]:
# выбросы для столбца rectal temperature
df[~df['rectal temperature'].between(lower_bound, upper_bound)]

,surgery,age,rectal temperature,pulse,respiratory rate,temperature of extremities,pain,outcome
5,2.0,1,NaN,NaN,NaN,2.0,2.0,1.0
7,1.0,1,NaN,60.0,NaN,3.0,NaN,2.0
8,2.0,1,NaN,80.0,36.0,3.0,4.0,3.0
16,1.0,9,NaN,128.0,36.0,3.0,4.0,2.0
20,1.0,1,39.9,72.0,60.0,1.0,5.0,1.0
...,...,...,...,...,...,...,...,...
282,1.0,1,NaN,70.0,16.0,3.0,2.0,2.0
288,1.0,1,NaN,NaN,NaN,NaN,NaN,1.0
293,1.0,1,NaN,78.0,24.0,3.0,NaN,3.0
295,1.0,1,NaN,120.0,70.0,4.0,2.0,3.0


In [ ]:
q1 = df['pulse'].quantile(0.25)
q3 = df['pulse'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
remove_outliers = df[df['pulse'].between(lower_bound, upper_bound)]
remove_outliers

,surgery,age,rectal temperature,pulse,respiratory rate,temperature of extremities,pain,outcome
0,2.0,1,38.5,66.0,28.0,3.0,5.0,2.0
1,1.0,1,39.2,88.0,20.0,NaN,3.0,3.0
2,2.0,1,38.3,40.0,24.0,1.0,3.0,1.0
4,2.0,1,37.3,104.0,35.0,NaN,NaN,2.0
6,1.0,1,37.9,48.0,16.0,1.0,3.0,1.0
...,...,...,...,...,...,...,...,...
295,1.0,1,NaN,120.0,70.0,4.0,2.0,3.0
296,2.0,1,37.2,72.0,24.0,3.0,4.0,3.0
297,1.0,1,37.5,72.0,30.0,4.0,4.0,2.0
298,1.0,1,36.5,100.0,24.0,3.0,3.0,1.0


In [ ]:
# выбросы для столбца pulse
df[~df['pulse'].between(lower_bound, upper_bound)]

,surgery,age,rectal temperature,pulse,respiratory rate,temperature of extremities,pain,outcome
3,1.0,9,39.1,164.0,84.0,4.0,2.0,2.0
5,2.0,1,NaN,NaN,NaN,2.0,2.0,1.0
28,1.0,1,NaN,NaN,NaN,NaN,NaN,2.0
41,2.0,9,39.0,150.0,72.0,NaN,NaN,1.0
52,2.0,1,NaN,NaN,NaN,1.0,1.0,1.0
55,1.0,9,38.6,160.0,20.0,3.0,3.0,2.0
56,1.0,1,NaN,NaN,NaN,NaN,NaN,1.0
58,1.0,1,NaN,NaN,20.0,4.0,5.0,2.0
74,1.0,9,NaN,NaN,NaN,NaN,NaN,2.0
78,1.0,1,NaN,NaN,NaN,3.0,5.0,2.0


In [ ]:
q1 = df['respiratory rate'].quantile(0.25)
q3 = df['respiratory rate'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
remove_outliers = df[df['respiratory rate'].between(lower_bound, upper_bound)]
remove_outliers

,surgery,age,rectal temperature,pulse,respiratory rate,temperature of extremities,pain,outcome
0,2.0,1,38.5,66.0,28.0,3.0,5.0,2.0
1,1.0,1,39.2,88.0,20.0,NaN,3.0,3.0
2,2.0,1,38.3,40.0,24.0,1.0,3.0,1.0
4,2.0,1,37.3,104.0,35.0,NaN,NaN,2.0
6,1.0,1,37.9,48.0,16.0,1.0,3.0,1.0
...,...,...,...,...,...,...,...,...
294,2.0,1,38.5,40.0,16.0,1.0,2.0,1.0
296,2.0,1,37.2,72.0,24.0,3.0,4.0,3.0
297,1.0,1,37.5,72.0,30.0,4.0,4.0,2.0
298,1.0,1,36.5,100.0,24.0,3.0,3.0,1.0


In [ ]:
# выбросы для столбца respiratory rate
df[~df['respiratory rate'].between(lower_bound, upper_bound)]

,surgery,age,rectal temperature,pulse,respiratory rate,temperature of extremities,pain,outcome
3,1.0,9,39.1,164.0,84.0,4.0,2.0,2.0
5,2.0,1,NaN,NaN,NaN,2.0,2.0,1.0
7,1.0,1,NaN,60.0,NaN,3.0,NaN,2.0
9,2.0,9,38.3,90.0,NaN,1.0,5.0,1.0
28,1.0,1,NaN,NaN,NaN,NaN,NaN,2.0
...,...,...,...,...,...,...,...,...
274,1.0,1,NaN,76.0,NaN,NaN,NaN,3.0
281,2.0,1,40.0,78.0,NaN,3.0,2.0,2.0
284,2.0,1,38.5,54.0,NaN,1.0,3.0,1.0
288,1.0,1,NaN,NaN,NaN,NaN,NaN,1.0


In [ ]:
q1 = df['temperature of extremities'].quantile(0.25)
q3 = df['temperature of extremities'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
remove_outliers = df[df['temperature of extremities'].between(lower_bound, upper_bound)]
remove_outliers

,surgery,age,rectal temperature,pulse,respiratory rate,temperature of extremities,pain,outcome
0,2.0,1,38.5,66.0,28.0,3.0,5.0,2.0
2,2.0,1,38.3,40.0,24.0,1.0,3.0,1.0
3,1.0,9,39.1,164.0,84.0,4.0,2.0,2.0
5,2.0,1,NaN,NaN,NaN,2.0,2.0,1.0
6,1.0,1,37.9,48.0,16.0,1.0,3.0,1.0
...,...,...,...,...,...,...,...,...
294,2.0,1,38.5,40.0,16.0,1.0,2.0,1.0
295,1.0,1,NaN,120.0,70.0,4.0,2.0,3.0
296,2.0,1,37.2,72.0,24.0,3.0,4.0,3.0
297,1.0,1,37.5,72.0,30.0,4.0,4.0,2.0


In [ ]:
# выбросы для столбца respiratory rate
df[~df['temperature of extremities'].between(lower_bound, upper_bound)]

,surgery,age,rectal temperature,pulse,respiratory rate,temperature of extremities,pain,outcome
1,1.0,1,39.2,88.0,20.0,NaN,3.0,3.0
4,2.0,1,37.3,104.0,35.0,NaN,NaN,2.0
17,2.0,1,37.5,48.0,24.0,NaN,NaN,1.0
23,1.0,9,38.3,130.0,60.0,NaN,2.0,1.0
25,2.0,1,37.8,60.0,42.0,NaN,NaN,1.0
28,1.0,1,NaN,NaN,NaN,NaN,NaN,2.0
36,2.0,1,38.3,112.0,16.0,NaN,NaN,3.0
37,1.0,1,37.8,72.0,NaN,NaN,5.0,1.0
39,1.0,9,39.2,146.0,96.0,NaN,NaN,2.0
41,2.0,9,39.0,150.0,72.0,NaN,NaN,1.0


In [ ]:
q1 = df['pain'].quantile(0.25)
q3 = df['pain'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
remove_outliers = df[df['pain'].between(lower_bound, upper_bound)]
remove_outliers

,surgery,age,rectal temperature,pulse,respiratory rate,temperature of extremities,pain,outcome
0,2.0,1,38.5,66.0,28.0,3.0,5.0,2.0
1,1.0,1,39.2,88.0,20.0,NaN,3.0,3.0
2,2.0,1,38.3,40.0,24.0,1.0,3.0,1.0
3,1.0,9,39.1,164.0,84.0,4.0,2.0,2.0
5,2.0,1,NaN,NaN,NaN,2.0,2.0,1.0
...,...,...,...,...,...,...,...,...
294,2.0,1,38.5,40.0,16.0,1.0,2.0,1.0
295,1.0,1,NaN,120.0,70.0,4.0,2.0,3.0
296,2.0,1,37.2,72.0,24.0,3.0,4.0,3.0
297,1.0,1,37.5,72.0,30.0,4.0,4.0,2.0


In [ ]:
# выбросы для столбца respiratory pain
df[~df['pain'].between(lower_bound, upper_bound)]

,surgery,age,rectal temperature,pulse,respiratory rate,temperature of extremities,pain,outcome
4,2.0,1,37.3,104.0,35.0,NaN,NaN,2.0
7,1.0,1,NaN,60.0,NaN,3.0,NaN,2.0
17,2.0,1,37.5,48.0,24.0,NaN,NaN,1.0
19,2.0,1,39.4,110.0,35.0,4.0,NaN,1.0
24,1.0,1,38.1,60.0,12.0,3.0,NaN,1.0
25,2.0,1,37.8,60.0,42.0,NaN,NaN,1.0
27,1.0,1,37.8,48.0,12.0,3.0,NaN,1.0
28,1.0,1,NaN,NaN,NaN,NaN,NaN,2.0
36,2.0,1,38.3,112.0,16.0,NaN,NaN,3.0
39,1.0,9,39.2,146.0,96.0,NaN,NaN,2.0


In [ ]:
q1 = df['outcome'].quantile(0.25)
q3 = df['outcome'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
remove_outliers = df[df['outcome'].between(lower_bound, upper_bound)]
remove_outliers

,surgery,age,rectal temperature,pulse,respiratory rate,temperature of extremities,pain,outcome
0,2.0,1,38.5,66.0,28.0,3.0,5.0,2.0
1,1.0,1,39.2,88.0,20.0,NaN,3.0,3.0
2,2.0,1,38.3,40.0,24.0,1.0,3.0,1.0
3,1.0,9,39.1,164.0,84.0,4.0,2.0,2.0
4,2.0,1,37.3,104.0,35.0,NaN,NaN,2.0
...,...,...,...,...,...,...,...,...
295,1.0,1,NaN,120.0,70.0,4.0,2.0,3.0
296,2.0,1,37.2,72.0,24.0,3.0,4.0,3.0
297,1.0,1,37.5,72.0,30.0,4.0,4.0,2.0
298,1.0,1,36.5,100.0,24.0,3.0,3.0,1.0


In [ ]:
# выбросы для столбца respiratory pain
df[~df['outcome'].between(lower_bound, upper_bound)]

,surgery,age,rectal temperature,pulse,respiratory rate,temperature of extremities,pain,outcome
132,NaN,1,38.0,48.0,20.0,3.0,4.0,NaN


Удаление строк с пропусками

In [ ]:
df.dropna().info()

<class 'pandas.core.frame.DataFrame'>
Index: 154 entries, 0 to 298
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery                     154 non-null    float64
 1   age                         154 non-null    int64  
 2   rectal temperature          154 non-null    float64
 3   pulse                       154 non-null    float64
 4   respiratory rate            154 non-null    float64
 5   temperature of extremities  154 non-null    float64
 6   pain                        154 non-null    float64
 7   outcome                     154 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 10.8 KB


Статистические данные до удаления пропусков в столбце age
1.64
1.0
1
2.173971905557624
4.726153846153846

In [ ]:
#статистисческие данные по столбцу age после удаления пропусков . Делаю вывод, что здесь MCAR. Распределение данных - нормальное
print(df.dropna()['age'].mean())
print(df.dropna()['age'].median())
print(df.dropna()['age'].mode()[0])

print(df.dropna()['age'].std())
print(df.dropna()['age'].var())

1.6233766233766234
1.0
1
2.1513878240907753
4.6284695696460405


Статистические данные до удаления пропусков в столбце surgery
1.3979933110367893
1.0
1.0
0.4903046443288189
0.24039864425040958

In [ ]:
#статистисческие данные по столбцу surgery после удаления пропусков . Делаю вывод, что здесь MCAR
print(df.dropna()['surgery'].mean())
print(df.dropna()['surgery'].median())
print(df.dropna()['surgery'].mode()[0])

print(df.dropna()['surgery'].std())
print(df.dropna()['surgery'].var())

1.4350649350649352
1.0
1.0
0.4973830164339389
0.24738986503692395


Статистические данные до удаления пропусков в столбце rectal temperature

38.16791666666667
38.2
38.0
0.7322886641121578
0.5362466875871686

In [ ]:
#статистисческие данные по столбцу rectal temperature после удаления пропусков . Делаю вывод, что здесь MCAR

print(df.dropna()['rectal temperature'].mean())
print(df.dropna()['rectal temperature'].median())
print(df.dropna()['rectal temperature'].mode()[0])

print(df.dropna()['rectal temperature'].std())
print(df.dropna()['rectal temperature'].var())

38.163636363636364
38.2
38.0
0.7318896825527826
0.5356625074272129


Статистические данные до удаления пропусков в столбце pulse

71.91304347826087
64.0
48.0
28.630556660735
819.7087747035573

In [ ]:
#статистисческие данные по столбцу pulse после удаления пропусков . Делаю вывод, что здесь НЕ MCAR

print(df.dropna()['pulse'].mean())
print(df.dropna()['pulse'].median())
print(df.dropna()['pulse'].mode()[0])

print(df.dropna()['pulse'].std())
print(df.dropna()['pulse'].var())

70.84415584415585
60.0
48.0
29.60968969429682
876.7337237925473


Статистические данные до удаления пропусков в столбце respiratory rate

30.417355371900825
24.5
20.0
17.64223138513467
311.2483282466308

In [ ]:
#статистисческие данные по столбцу respiratory rate после удаления пропусков . Делаю вывод, что здесь НЕ MCAR

print(df.dropna()['respiratory rate'].mean())
print(df.dropna()['respiratory rate'].median())
print(df.dropna()['respiratory rate'].mode()[0])

print(df.dropna()['respiratory rate'].std())
print(df.dropna()['respiratory rate'].var())

29.649350649350648
27.0
16.0
16.122299122251533
259.92852898735254


Статистические данные до удаления пропусков в столбце temperature of extremities

2.348360655737705
3.0
3.0
1.0450536920112763
1.0921372191863994

In [ ]:
#статистисческие данные по столбцу temperature of extremities после удаления пропусков . Делаю вывод, что здесь MCAR

print(df.dropna()['temperature of extremities'].mean())
print(df.dropna()['temperature of extremities'].median())
print(df.dropna()['temperature of extremities'].mode()[0])

print(df.dropna()['temperature of extremities'].std())
print(df.dropna()['temperature of extremities'].var())

2.220779220779221
2.5
3.0
1.0617949609850506
1.127408539173245


Статистические данные до удаления пропусков в столбце pain

2.951020408163265
3.0
3.0
1.3079395711243496
1.7107059217129477

In [ ]:
#статистисческие данные по столбцу pain после удаления пропусков . Делаю вывод, что здесь MCAR

print(df.dropna()['pain'].mean())
print(df.dropna()['pain'].median())
print(df.dropna()['pain'].mode()[0])

print(df.dropna()['pain'].std())
print(df.dropna()['pain'].var())

2.6948051948051948
3.0
3.0
1.2171321810373057
1.4814107461166288


Статистические данные до удаления пропусков в столбце outcome

1.5518394648829432
1.0
1.0
0.7371869448054943
0.5434445915916589

In [ ]:
#статистисческие данные по столбцу pain после удаления пропусков . Делаю вывод, что здесь MCAR

print(df.dropna()['outcome'].mean())
print(df.dropna()['outcome'].median())
print(df.dropna()['outcome'].mode()[0])

print(df.dropna()['outcome'].std())
print(df.dropna()['outcome'].var())

1.4805194805194806
1.0
1.0
0.6975285898118851
0.48654613360495713


Сделаем замену пропусков средним только у тех столбцов, у которых произошло значительное изменение статистических данных после удаления пропусков : pulse, respiratory rate

In [ ]:
#столбец pulse
fill_mean = df.copy()

fill_mean['pulse'] = fill_mean['pulse'].fillna(df['pulse'].mean())

In [ ]:
#после замены средним у нас есть значительное изменение только в 2 квартиле
print(df['pulse'].describe())
print('-----------------------------------')
print(fill_mean['pulse'].describe())

count    276.000000
mean      71.913043
std       28.630557
min       30.000000
25%       48.000000
50%       64.000000
75%       88.000000
max      184.000000
Name: pulse, dtype: float64
-----------------------------------
count    300.000000
mean      71.913043
std       27.457472
min       30.000000
25%       48.000000
50%       68.000000
75%       88.000000
max      184.000000
Name: pulse, dtype: float64


In [ ]:
#столбец respiratory rate
fill_mean = df.copy()

fill_mean['respiratory rate'] = fill_mean['respiratory rate'].fillna(df['respiratory rate'].mean())

In [ ]:
#после замены средним у нас есть значительное изменение только в 2 квартиле
print(df['respiratory rate'].describe())
print('-----------------------------------')
print(fill_mean['respiratory rate'].describe())

count    242.000000
mean      30.417355
std       17.642231
min        8.000000
25%       18.500000
50%       24.500000
75%       36.000000
max       96.000000
Name: respiratory rate, dtype: float64
-----------------------------------
count    300.000000
mean      30.417355
std       15.838952
min        8.000000
25%       20.000000
50%       30.000000
75%       34.250000
max       96.000000
Name: respiratory rate, dtype: float64


Сделаем замену пропусков модой только у тех столбцов, у которых произошло значительное изменение статистических данных после удаления пропусков : pulse, respiratory rate

In [ ]:
#столбец pulse
df['pulse'].value_counts()

pulse
48.0     28
60.0     25
40.0     18
52.0     12
88.0     12
44.0     11
100.0    11
72.0     11
120.0    10
42.0     10
64.0      8
96.0      8
84.0      8
80.0      8
54.0      7
56.0      7
50.0      6
66.0      6
104.0     5
68.0      5
92.0      4
70.0      4
76.0      4
108.0     3
86.0      3
78.0      3
90.0      3
36.0      3
112.0     3
114.0     2
45.0      2
140.0     2
124.0     2
150.0     2
130.0     2
30.0      2
132.0     1
129.0     1
184.0     1
75.0      1
65.0      1
164.0     1
136.0     1
49.0      1
98.0      1
38.0      1
160.0     1
46.0      1
128.0     1
146.0     1
110.0     1
82.0      1
Name: count, dtype: int64

In [ ]:
df_fill_mode = df.copy()
df_fill_mode['pulse'] = df_fill_mode['pulse'].fillna(df['pulse'].mode()[0])

df_fill_mode['pulse'].value_counts()

pulse
48.0     52
60.0     25
40.0     18
52.0     12
88.0     12
44.0     11
100.0    11
72.0     11
120.0    10
42.0     10
64.0      8
96.0      8
84.0      8
80.0      8
54.0      7
56.0      7
50.0      6
66.0      6
104.0     5
68.0      5
92.0      4
70.0      4
76.0      4
108.0     3
86.0      3
78.0      3
90.0      3
36.0      3
112.0     3
114.0     2
45.0      2
140.0     2
124.0     2
150.0     2
130.0     2
30.0      2
132.0     1
129.0     1
184.0     1
75.0      1
65.0      1
164.0     1
136.0     1
49.0      1
98.0      1
38.0      1
160.0     1
46.0      1
128.0     1
146.0     1
110.0     1
82.0      1
Name: count, dtype: int64

In [ ]:
#71.91304347826087
#64.0
#48.0
#28.630556660735
#819.7087747035573
#Делаю вывод, что замена пропусков модой можно сделать

print(df.dropna()['pulse'].mean())
print(df.dropna()['pulse'].median())
print(df.dropna()['pulse'].mode()[0])

print(df.dropna()['pulse'].std())
print(df.dropna()['pulse'].var())

70.84415584415585
60.0
48.0
29.60968969429682
876.7337237925473


In [ ]:
#столбец respiratory rate
df['respiratory rate'].value_counts()

respiratory rate
20.0    28
24.0    27
16.0    22
12.0    19
30.0    19
40.0    17
36.0    16
28.0    13
32.0    11
18.0     8
48.0     6
60.0     4
14.0     4
42.0     3
68.0     3
10.0     3
44.0     3
35.0     3
80.0     3
9.0      2
51.0     2
90.0     2
70.0     2
22.0     2
96.0     2
21.0     2
84.0     2
50.0     2
15.0     1
72.0     1
66.0     1
34.0     1
52.0     1
8.0      1
23.0     1
58.0     1
13.0     1
26.0     1
88.0     1
25.0     1
Name: count, dtype: int64

In [ ]:
df_fill_mode = df.copy()
df_fill_mode['respiratory rate'] = df_fill_mode['respiratory rate'].fillna(df['respiratory rate'].mode()[0])

df_fill_mode['respiratory rate'].value_counts()

respiratory rate
20.0    86
24.0    27
16.0    22
12.0    19
30.0    19
40.0    17
36.0    16
28.0    13
32.0    11
18.0     8
48.0     6
60.0     4
14.0     4
42.0     3
68.0     3
10.0     3
44.0     3
35.0     3
80.0     3
9.0      2
51.0     2
90.0     2
70.0     2
22.0     2
96.0     2
21.0     2
84.0     2
50.0     2
15.0     1
72.0     1
66.0     1
34.0     1
52.0     1
8.0      1
23.0     1
58.0     1
13.0     1
26.0     1
88.0     1
25.0     1
Name: count, dtype: int64

In [ ]:
#Исходные данные
#30.417355371900825
#24.5
#20.0
#17.64223138513467
#311.2483282466308
#Делаю вывод, что замена пропусков модой можно сделать.

print(df.dropna()['respiratory rate'].mean())
print(df.dropna()['respiratory rate'].median())
print(df.dropna()['respiratory rate'].mode()[0])

print(df.dropna()['respiratory rate'].std())
print(df.dropna()['respiratory rate'].var())

29.649350649350648
27.0
16.0
16.122299122251533
259.92852898735254


Замену пропусков следующими/предыдущими значениями делать не стала, так как столбцы pulse и respiratory rate не являются временными рядами.